bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/handler/ICalendarHandler.java

```diff
@@ -0,0 +1,392 @@
+            Channel currentEventEndChannel = this.thing.getChannel(CHANNEL_CURRENT_EVENT_END);
+            Channel nextEventTitleChannel = this.thing.getChannel(CHANNEL_NEXT_EVENT_TITLE);
+            Channel nextEventStartChannel = this.thing.getChannel(CHANNEL_NEXT_EVENT_START);
+            Channel nextEventEndChannel = this.thing.getChannel(CHANNEL_NEXT_EVENT_END);
+            if (eventPresenceChannel == null || currentEventTitleChannel == null || currentEventStartChannel == null
+                    || currentEventEndChannel == null || nextEventTitleChannel == null || nextEventStartChannel == null
+                    || nextEventEndChannel == null) {
+                logger.warn("Could not retrieve one or more calendar channels. Not updating.");
+                return;
+            }
+
+            Instant now = Instant.now();
+            if (calendar.isEventPresent(now)) {
+                this.updateState(eventPresenceChannel.getUID(), OnOffType.ON);
+                Event currentEvent = calendar.getCurrentEvent(now);
+                if (currentEvent == null) {
+                    logger.warn("Unexpected inconsistency of internal API. Not Updating event details.");
+                } else {
+                    this.updateState(currentEventTitleChannel.getUID(), new StringType(currentEvent.title));
+                    this.updateState(currentEventStartChannel.getUID(),
+                            new DateTimeType(currentEvent.start.atZone(ZoneId.systemDefault())));
+                    this.updateState(currentEventEndChannel.getUID(),
+                            new DateTimeType(currentEvent.end.atZone(ZoneId.systemDefault())));
+                }
+            } else {
+                this.updateState(eventPresenceChannel.getUID(), OnOffType.OFF);
+                this.updateState(currentEventTitleChannel.getUID(), UnDefType.NULL);
+                this.updateState(currentEventStartChannel.getUID(), UnDefType.NULL);
+                this.updateState(currentEventEndChannel.getUID(), UnDefType.NULL);
+            }
+
+            Event nextEvent = calendar.getNextEvent(now);
+            if (nextEvent != null) {
+                this.updateState(nextEventTitleChannel.getUID(), new StringType(nextEvent.title));
+                this.updateState(nextEventStartChannel.getUID(),
+                        new DateTimeType(nextEvent.start.atZone(ZoneId.systemDefault())));
+                this.updateState(nextEventEndChannel.getUID(),
+                        new DateTimeType(nextEvent.end.atZone(ZoneId.systemDefault())));
+            } else {
+                this.updateState(nextEventTitleChannel.getUID(), UnDefType.NULL);
+                this.updateState(nextEventStartChannel.getUID(), UnDefType.NULL);
+                this.updateState(nextEventEndChannel.getUID(), UnDefType.NULL);
+            }
+
+            // FIRSTLY: process all Command Tags in all Calendar Events which ENDED since updateStates was last called
```

Comment:
IMHO the comment IS necessary; you need to first process the just ended events before secondly processing the just begun events! If it would be done the other way round, then the functionality fails.

Reply:
Don